gnovm/pkg/gnolang/nodes.go

```diff
@@ -1146,6 +1146,7 @@ func PrecompileMemPackage(memPkg *std.MemPackage) error {
 }
 
 // Returns the code fileset minus any spurious or test files.
+// Here?
 func ParseMemPackage(memPkg *std.MemPackage) (fset *FileSet) {
```

Comment:
I believe it's the ideal location for early arrivals, package uploads, and early cleanups. @piux2, could you please verify?

Reply:
In the Gno language environment, ParseMemPackage() in gnolang/nodes.go is called in six distinct scenarios, each with different contexts. These scenarios can be divided into three main categories: blockchain application, file testing, and local development.

### Blockchain Application:
This involves the Gno language as an application running on the GnoLand blockchain. There are two specific scenarios under this category:

1. `gno.land start`
  When this command is executed, the Gno virtual machine (VM) only reads the gnovm/stdlibs contracts from the local file system.
  Instead of loading everything into memory, it initializes a package getter and loads packages during runtime as needed.

2. `gnokey maketx addpkg`
   In this case, the VM parses contracts loaded from the GnoLand app store on-chain.
All non-stdlib contracts (identified as "p" and "r") in the Gno app store are derived from the Tendermint Consensus and genesis transactions.

### File Testing:

This scenario involves loading packages into a machine instance from a testing store, without involving blockchain applications. Packages can be loaded from the local file system. 

3. `go test ./tests -run '^TestFiles$'`

import.TestStore() is used.

### Gno Local Development and Test:

4. `gno run` imports.TestStore() is used.
5. `gno repl` imports.TestStore() is used.
   
   
6. `gno test ./examples`


Gno modules (Gno mod) are used to facilitate local development and testing. They are not entities on the blockchain.
The goal is to avoid mixing the logic of on-chain production, local development, and testing within the gnolang/nodes.go functions. Each scenario handles the parsing and loading of packages differently. therefore the logic should be implemted outside of the pure gnolang function in a language pkg for the specific needs.  

```mermaid
graph TD
    A[1 'gno.land start'] -->AB[makeGenesis] -->B[NewApp]
    B --> C[VMKeeper.Initialize]
    C --> D[m.PreprocessAllFilesAndSaveBlockNodes]
    D --> E[gnolang/nodes.go: ParseMemPackage]

    F[2 'gnokey maketx addpkg'] -->|Network| H[tm2.PRC]
    H --> I[vmHandler]
    I --> J[VMKeeper.AddPackage]
    J --> U[m.RunMemPackage] 

    K[3 'go test ./tests -run '^TestFiles$''] --> L[RunFileTest]
    L --> U 

    M[4 'gno run'] --> N[import.TestStore]
    N --> U 

    O[5 'gno repl'] --> P[import.TestStore]
    P --> U 

    Q['6 gno test ./examples'] --> R[execTest]
    R -->S[gnoTestPkg]
    S --> U 
    U --> E

     style A stroke:#1548c6,stroke-width:2px
     style F stroke:#1548c6,stroke-width:2px
     style K stroke:#111111,stroke-width:2px
     style M stroke:#111111,stroke-width:2px
     style O stroke:#111111,stroke-width:2px
     style Q stroke:#111111,stroke-width:2px
   

     style E stroke:#b548c6,stroke-width:2px
     style U stroke:#b548c6,stroke-width:2px
     style D stroke:#b548c6,stroke-width:2px
     
 
```